### PR TITLE
Deal with node balance underflow

### PR DESCRIPTION
**Pull request by potuz**
In the event that an underflow on a forkchoice node occurs (in the doubly linked tree implementation), set the node balance to zero instead of returning an error. This mimics the behavior on the protoarray implementation. 

This PR also adds a debug log to help find the cause for these underflows.

### CODE REVIEW

File: beacon-chain/forkchoice/doubly-linked-tree/forkchoice.go [hunk 1]
```diff
@@ -249,9 +252,21 @@ func (f *ForkChoice) updateBalances(newBalances []uint64) error {
...
+					f.store.proposerBoostLock.RLock()
+					log.WithFields(logrus.Fields{
+						"NodeRoot":                   fmt.Sprintf("%#x", bytesutil.Trunc(vote.currentRoot[:])),
+						"OldBalance":                 oldBalance,
+						"NodeBalance":                currentNode.balance,
+						"NodeWeight":                 currentNode.weight,
+						"ProposerBoostRoot":          f.store.proposerBoostRoot,
+						"PreviousProposerBoostRoot":  f.store.previousProposerBoostRoot,
+						"PreviousProposerBoostScore": f.store.previousProposerBoostScore,
+					}).Debug("node with invalid balance, setting it to zero")
```
Comment (terencechain) [inline beacon-chain/forkchoice/doubly-linked-tree/forkchoice.go:264]: I don't think this log is necessary. What action do you expect users to take from it? If we make a case for this, there are many more edge cases to log around fork choice.

If you feel strongly about adding this, please make sure the log is in the proper format. e.g., not cap and root is formatted

**Review by terencechain — APPROVED** [review]

File: beacon-chain/forkchoice/doubly-linked-tree/forkchoice.go [hunk 2]
```diff
@@ -250,9 +253,21 @@ func (f *ForkChoice) updateBalances(newBalances []uint64) error {
...
+					f.store.proposerBoostLock.RLock()
+					log.WithFields(logrus.Fields{
+						"nodeRoot":                   fmt.Sprintf("%#x", bytesutil.Trunc(vote.currentRoot[:])),
+						"oldBalance":                 oldBalance,
+						"nodeBalance":                currentNode.balance,
+						"nodeWeight":                 currentNode.weight,
+						"proposerBoostRoot":          fmt.Sprintf("%#x", bytesutil.Trunc(f.store.proposerBoostRoot[:])),
+						"previousProposerBoostRoot":  fmt.Sprintf("%#x", bytesutil.Trunc(f.store.previousProposerBoostRoot[:])),
+						"previousProposerBoostScore": f.store.previousProposerBoostScore,
+					}).Debug("node with invalid balance, setting it to zero")
```
Comment (prestonvanloon) [inline beacon-chain/forkchoice/doubly-linked-tree/forkchoice.go:265]: Can this be a warning level message for now? In theory, it should never happen so we should be alerted to when it does happen.

File: beacon-chain/forkchoice/protoarray/store.go [hunk 3]
```diff
@@ -426,6 +428,16 @@ func (s *Store) applyWeightChanges(
...
 			if n.weight < d {
+				s.proposerBoostLock.RLock()
+				log.WithFields(logrus.Fields{
+					"nodeDelta":                  d,
+					"nodeRoot":                   fmt.Sprintf("%#x", bytesutil.Trunc(n.root[:])),
+					"nodeWeight":                 n.weight,
+					"proposerBoostRoot":          fmt.Sprintf("%#x", bytesutil.Trunc(s.proposerBoostRoot[:])),
+					"previousProposerBoostRoot":  fmt.Sprintf("%#x", bytesutil.Trunc(s.previousProposerBoostRoot[:])),
+					"previousProposerBoostScore": s.previousProposerBoostScore,
+				}).Debug("node with invalid weight, setting it to zero")
```
Comment (prestonvanloon) [inline beacon-chain/forkchoice/protoarray/store.go:439]: Same here. I think WARN is appropriate.

File: beacon-chain/forkchoice/doubly-linked-tree/forkchoice.go (hunk 2 above)
Comment (prestonvanloon) [inline beacon-chain/forkchoice/doubly-linked-tree/forkchoice.go:265]:
```suggestion
					}).Warning("node with invalid balance, setting it to zero")
```

File: beacon-chain/forkchoice/protoarray/store.go (hunk 3 above)
Comment (prestonvanloon) [inline beacon-chain/forkchoice/protoarray/store.go:439]:
```suggestion
				}).Warning("node with invalid weight, setting it to zero")
```

**Review by prestonvanloon — APPROVED** [review]

**Pull request merged**